File: lib/pure/crypto/sign.ex
defmodule Meta do
	defstruct sig: nil, sig2: nil, revealed: []
end

defmodule CryptoSign do
  defstruct data: nil, meta: %Meta{} #sig and sig2 go in meta
  def params do
    :crypto.ec_curve(:secp256k1)
  end
	def shared_secret(pub, priv) do
		:crypto.compute_key(:ecdh, elem(Base.decode64(pub), 1), elem(Base.decode64(priv), 1), params)
	end
  def new_key do
    {pub, priv} = :crypto.generate_key(:ecdh, params)
    pub = Base.encode64(pub)
    priv = Base.encode64(priv)
    {pub, priv}
  end
  def sign(s, priv) do
    {:ok, priv} = Base.decode64(priv)
    :crypto.sign(:ecdsa, :sha256, s, [priv, params]) 
    |> Base.encode64
  end
  def verify(s, sig, pub) do
    {x, sig} = Base.decode64(sig)
    {y, pub} = Base.decode64(pub)
		cond do
			x != :ok -> false
			y != :ok -> false
			true -> :crypto.verify(:ecdsa, :sha256, s, sig, [pub, params])
		end
  end
  def sign_tx(tx, pub, priv) do
		if :__struct__ in Map.keys(tx) and tx.__struct__ == :Elixir.CryptoSign do
			cb = tx.data
			m = tx.meta
		else
			cb = tx
			m = %Meta{}
		end
		h = DetHash.doit(cb)
		sig = sign(h, priv)
		a = true
		if cb.pub == pub do
				m = %{m | sig: sig}
				a = false
		end
		if :pub2 in Map.keys(cb) and (cb.pub2 == pub) do
				m = %{m | sig2: sig}
				a = false
		end
		if a do IO.puts("not for me to sign") end
		%CryptoSign{meta: m, data: cb}
  end
  def verify_tx(tx) do
    h = DetHash.doit(tx.data)
		pub = tx.data.pub
    verify(h, tx.meta.sig, pub)
  end
  def test do
    {pub, priv} = new_key
    tx = %{pub: "BCmhaRq42NNQe6ZpRHIvDxHBThEE3LDBN68KUWXmCTKUZvMI8Ol1g9yvDVTvMsZbqHQZ5j8E7sKVCgZMJR7lQWc="}
    tx = sign_tx(tx, pub, priv) 
    true = verify_tx(tx)
    s = "test string"
    sig = sign(s, priv)
    true = verify(s, sig, pub)
    false = verify(s <> " ", sig, pub)
    "success"
  end
	def check_sig2(tx) do
		h = DetHash.doit(tx.data)
		pub = tx.data.pub2
		verify(h, tx.meta.sig2, pub)
	end
end

File: src/tree/sign.erl
-module(sign).
-export([test/0,new_key/0,sign_tx/4,sign/2,verify_sig/3,shared_secret/2,verify/2,data/1,revealed/1,empty/1,empty/0]).
-record(signed, {data="", sig="", sig2="", revealed=[]}).
empty() -> #signed{}.
empty(X) -> #signed{data=X}.
data(X) -> X#signed.data.
revealed(X) -> X#signed.revealed.
en(X) -> base64:encode(X).
de(X) -> base64:decode(X).
params() -> crypto:ec_curve(secp256k1).
shared_secret(Pub, Priv) -> en(crypto:compute_key(ecdh, de(Pub), de(Priv), params())).
new_key() -> 
    {Pub, Priv} = crypto:generate_key(ecdh, params()),
    {en(Pub), en(Priv)}.
sign(S, Priv) -> en(crypto:sign(ecdsa, sha256, term_to_binary(S), [de(Priv), params()])).
verify_sig(S, Sig, Pub) -> crypto:verify(ecdsa, sha256, term_to_binary(S), de(Sig), [de(Pub), params()]).
verify_1(Tx, Pub) -> verify_sig(Tx#signed.data, Tx#signed.sig, Pub).
verify_2(Tx, Pub) -> verify_sig(Tx#signed.data, Tx#signed.sig2, Pub).
verify_both(Tx, Pub1, Pub2) ->
    X = verify_1(Tx, Pub1),
    Y = verify_2(Tx, Pub1),
    if
        X -> verify_2(Tx, Pub2);
        Y -> verify_1(Tx, Pub2);
        true -> false
    end.
verify(SignedTx, Accounts) ->
    Tx = SignedTx#signed.data,
    N1 = element(2, Tx),
    Acc1 = block_tree:account(N1, Accounts),
    Type = element(1, Tx),
    if
	(Type == channel_block) or (Type == tc) ->
	    N2 = element(3, Tx),
	    Acc2 = block_tree:account(N2, Accounts),
	    verify_both(SignedTx, accounts:pub(Acc1), accounts:pub(Acc2));
	true -> verify_1(SignedTx, accounts:pub(Acc1))
    end.
sign_tx(SignedTx, Pub, Priv, Accounts) when element(1, SignedTx) == signed ->
    Tx = SignedTx#signed.data,
    R = SignedTx#signed.revealed,
    Sig = sign(Tx, Priv),
    N = element(2, Tx),
    Acc = block_tree:account(N, Accounts),
    APub = accounts:pub(Acc),
    if
	APub == Pub -> #signed{data=Tx, sig=Sig, sig2=SignedTx#signed.sig2, revealed=R};
	true ->
	    N2 = element(3, Tx),
	    Acc2 = block_tree:account(N2, Accounts),
	    BPub = accounts:pub(Acc2),
	    Pub = BPub,
	    #signed{data=Tx, sig=SignedTx#signed.sig, sig2=Sig, revealed=R}
    end;
sign_tx(Tx, Pub, Priv, Accounts) ->
    Sig = sign(Tx, Priv),
    N = element(2, Tx),
    Acc = block_tree:account(N, Accounts),
    APub = accounts:pub(Acc),
    if
	APub == Pub -> #signed{data=Tx, sig=Sig};
	true ->
	    N2 = element(3, Tx),
	    Acc2 = block_tree:account(N2, Accounts),
	    Pub = accounts:pub(Acc2),
	    #signed{data=Tx, sig2=Sig}
    end.

test() ->
    {Pub, Priv} = new_key(),
    {Pub2, Priv2} = new_key(),
    Acc = accounts:empty(Pub),
    Acc2 = accounts:empty(Pub2),
    Accounts = dict:store(1, Acc2, dict:store(0, Acc, dict:new())),
    Tx = {channel_block, 0, 1},
    Signed = sign_tx(sign_tx(Tx, Pub, Priv, Accounts), Pub2, Priv2, Accounts),
    Signed2 = sign_tx({spend, 0}, Pub, Priv, Accounts),
    true = verify(Signed2, Accounts),
    true = verify(Signed, Accounts),
    true = verify_both(Signed, Pub2, Pub) 
        and (verify_both(Signed, Pub, Pub2)
        and not verify_both(Signed, Pub, Pub)),
    success.
